File: world/wrappers/record_episode.py
"""world/wrappers/record_episode.py - Records episode statistics."""
from __future__ import annotations
from typing import Any, Dict, List, Optional, Tuple
from ..core import Env, StepResult


class RecordEpisodeWrapper(Env):
    """Collects per-episode statistics (length, cumulative reward).

    Access via ``wrapper.episode_stats`` after episodes complete.
    """

    def __init__(self, env: Env):
        self._env = env
        self.episode_stats: List[Dict[str, float]] = []
        self._ep_reward = 0.0
        self._ep_length = 0

    @property
    def observation_space(self):
        return self._env.observation_space

    @property
    def action_space(self):
        return self._env.action_space

    def reset(self, *, seed=None, options=None) -> Tuple[Any, Dict]:
        self._ep_reward = 0.0
        self._ep_length = 0
        return self._env.reset(seed=seed, options=options)

    def step(self, action) -> StepResult:
        result = self._env.step(action)
        self._ep_reward += result.reward
        self._ep_length += 1
        if result.done:
            self.episode_stats.append({
                "reward": self._ep_reward,
                "length": self._ep_length,
                "terminated": result.terminated,
                "truncated": result.truncated,
            })
        return result

    def render(self, **kw):
        return self._env.render(**kw)

    def close(self):
        self._env.close()

    def summary(self) -> Dict[str, float]:
        if not self.episode_stats:
            return {}
        rewards = [e["reward"] for e in self.episode_stats]
        lengths = [e["length"] for e in self.episode_stats]
        import statistics
        return {
            "episodes": len(rewards),
            "mean_reward": statistics.mean(rewards),
            "max_reward": max(rewards),
            "min_reward": min(rewards),
            "mean_length": statistics.mean(lengths),
        }

    def __repr__(self):
        return f"RecordEpisodeWrapper({self._env})"

File: world/wrappers/record_episode.ts
import { Env, StepResult } from '../core';

export class RecordEpisodeWrapper<ObsType, ActType> extends Env<ObsType, ActType> {
  private _env: Env<ObsType, ActType>;
  episode_reward = 0;
  episode_length = 0;
  episode_stats: { reward: number; length: number }[] = [];
  current_episode: { rewards: number[]; lengths: number[] } = { rewards: [], lengths: [] };

  get observationSpace() { return this._env.observationSpace; }
  get actionSpace() { return this._env.actionSpace; }

  constructor(env: Env<ObsType, ActType>) {
    super();
    this._env = env;
  }

  reset(options?: { seed?: number }) {
    this.episode_reward = 0;
    this.episode_length = 0;
    this.current_episode = { rewards: [], lengths: [] };
    return this._env.reset(options);
  }

  step(action: ActType): StepResult<ObsType> {
    const result = this._env.step(action);
    this.episode_reward += result.reward;
    this.episode_length++;
    this.current_episode.rewards.push(result.reward);
    this.current_episode.lengths.push(this.episode_length);

    if (result.done) {
      this.episode_stats.push({
        reward: this.episode_reward,
        length: this.episode_length,
      });
    }
    return result;
  }

  summary(): { episodes: number; mean_reward: number; max_reward: number; mean_length: number } {
    const stats = this.episode_stats;
    if (stats.length === 0) {
      return { episodes: 0, mean_reward: 0, max_reward: 0, mean_length: 0 };
    }
    const rewards = stats.map(s => s.reward);
    const lengths = stats.map(s => s.length);
    return {
      episodes: stats.length,
      mean_reward: rewards.reduce((a, b) => a + b, 0) / rewards.length,
      max_reward: Math.max(...rewards),
      mean_length: lengths.reduce((a, b) => a + b, 0) / lengths.length,
    };
  }

  render() { return this._env.render(); }
  close() { this._env.close(); }
}